File: leetcode_75/Binary_Tree_DFS/cpp/1448.cpp
#include <algorithm>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int goodNodes(TreeNode *root)
    {
        this->count_good_nodes(root, -100000);

        return this->count;
    }

private:
    int count = 0;

    int count_good_nodes(TreeNode *node, int mx)
    {
        if (node == nullptr)
        {
            return 0;
        }

        this->count += node->val >= mx ? 1 : 0;
        mx = max(mx, node->val);

        return this->count_good_nodes(node->left, mx) + this->count_good_nodes(node->right, mx);
    }
};

File: leetcode_75/Binary_Tree_DFS/python/1448.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    count = 0

    def count_good_nodes(self, node: TreeNode, mx):

        if node is None:
            return 0

        self.count += int(node.val >= mx)
        mx = max(mx, node.val)

        return self.count_good_nodes(node.left, mx) + self.count_good_nodes(node.right, mx)

    def goodNodes(self, root: TreeNode) -> int:
        self.count_good_nodes(root, -10 ** 5)
        return self.count
